In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
# read and removal of all empty value
from csv import reader

groceries = []
with open(r'Dataset_Day15.csv') as csvfile:
    read = reader(csvfile)
    for row in read:
        # Remove 'NaN' and empty strings
        row_new = [x for x in row if x != 'NaN' and x.strip() != '']
        groceries.append(row_new)

In [3]:
groceries[0:5]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']]

In [5]:
!pip install mlxtend

In [6]:
#Transform the data using TransactonEncoder to perform Market Basket Analysis.
from mlxtend.preprocessing import TransactionEncoder

encoder = TransactionEncoder()

transactions = encoder.fit(groceries).transform(groceries)

import pandas as pd

itemsets = pd.DataFrame(transactions, columns = encoder.columns_)
display(itemsets.head())
display(itemsets.info())

,asparagus,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,black tea,blueberries,...,turkey,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,True,True,False,True,False,False,False,False,False,...,False,True,False,False,True,False,False,True,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,True,False,False,False,False,False,...,True,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Columns: 120 entries,  asparagus to zucchini
dtypes: bool(120)
memory usage: 879.2 KB


None

In [7]:
#Use min_support = 0.02 to find frequent itemsets.
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(itemsets, min_support = 0.02, use_colnames = True)
frequent_itemsets

,support,itemsets
0,0.020397,(almonds)
1,0.033329,(avocado)
2,0.033729,(brownies)
3,0.087188,(burgers)
4,0.030129,(butter)
...,...,...
98,0.020131,"(mineral water, whole wheat rice)"
99,0.022930,"(olive oil, spaghetti)"
100,0.025197,"(pancakes, spaghetti)"
101,0.021197,"(spaghetti, shrimp)"


In [8]:
frequent_itemsets['length'] = frequent_itemsets['itemsets'].str.len()
frequent_itemsets.head(-5)

,support,itemsets,length
0,0.020397,(almonds),1
1,0.033329,(avocado),1
2,0.033729,(brownies),1
3,0.087188,(burgers),1
4,0.030129,(butter),1
...,...,...,...
93,0.033729,"(mineral water, pancakes)",2
94,0.023597,"(mineral water, shrimp)",2
95,0.023064,"(mineral water, soup)",2
96,0.059725,"(mineral water, spaghetti)",2


Observations:-

Frequent Single Items: Items like burgers (8.7%), spaghetti (not shown in full list but likely high), and brownies (3.37%) are among the most frequent individual products. Even relatively less common items like almonds (2.03%) and butter (3.01%) meet the minimum support of 0.02, thus getting included.
Frequent Pairs (Length = 2): Common pairs include “spaghetti & mineral water” (5.97%), “pancakes & mineral water” (3.37%), and “shrimp & mineral water” (2.36%).
Mineral Water: It appears to be a key product it several purchases as a highly associated item.
Itemset Size: In this table, we only see itemsets of sizes 1 and 2, it is likely that any larger itemsets are way less common so it falls below the frequency threshold.
Strategy: From this information, we can improve the way the supermarkets place their products in their shelves, strategically placing the common pairs of items close to each other.

In [9]:
#Find top 5 antecedent -> consequent rules based on ‘conviction’, ‘leverage’, ‘lift’. Explain your findings.
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.15)
rules

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(eggs),(burgers),0.179709,0.087188,0.028796,0.160237,1.837830,1.0,0.013128,1.086988,0.555754,0.120941,0.080026,0.245256
1,(burgers),(eggs),0.087188,0.179709,0.028796,0.330275,1.837830,1.0,0.013128,1.224818,0.499424,0.120941,0.183552,0.245256
2,(burgers),(french fries),0.087188,0.170911,0.021997,0.252294,1.476173,1.0,0.007096,1.108844,0.353384,0.093168,0.098160,0.190499
3,(burgers),(mineral water),0.087188,0.238368,0.024397,0.279817,1.173883,1.0,0.003614,1.057552,0.162275,0.081009,0.054420,0.191083
4,(burgers),(spaghetti),0.087188,0.174110,0.021464,0.246177,1.413918,1.0,0.006283,1.095602,0.320707,0.089494,0.087260,0.184727
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
66,(whole wheat rice),(mineral water),0.058526,0.238368,0.020131,0.343964,1.442993,1.0,0.006180,1.160960,0.326080,0.072736,0.138644,0.214208
67,(olive oil),(spaghetti),0.065858,0.174110,0.022930,0.348178,1.999758,1.0,0.011464,1.267048,0.535186,0.105651,0.210764,0.239939
68,(pancakes),(spaghetti),0.095054,0.174110,0.025197,0.265077,1.522468,1.0,0.008647,1.123778,0.379218,0.103279,0.110144,0.204897
69,(shrimp),(spaghetti),0.071457,0.174110,0.021197,0.296642,1.703760,1.0,0.008756,1.174209,0.444850,0.094474,0.148363,0.209194


Observations:-

Olive Oil → Spaghetti has highest lift (about 1.999) and conviction (about 1.267), meaning olive oil is a great predictor of spaghetti purchases.
Rules like Burgers → Eggs and Burgers → French Fries have conviction above 1.22 and lift above 1.83, this means burgers can strongly predict the purchases for Eggs and French Fries.
Rules like Shrimp → Spaghetti or Pancakes → Spaghetti have lower leverages each, but they show significant lift. This suggests that while these combinations are unconventional, they are supported by consistent customers.

In [10]:
# Top 5 rules by conviction
top5_conviction = rules.sort_values('conviction', ascending=False).head(5)
# Top 5 rules by leverage
top5_leverage = rules.sort_values('leverage', ascending=False).head(5)
# Top 5 rules by lift
top5_lift = rules.sort_values('lift', ascending=False).head(5)
top5_conviction

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
62,(soup),(mineral water),0.050527,0.238368,0.023064,0.456464,1.914955,1.0,0.011020,1.401255,0.503221,0.086760,0.286354,0.276610
53,(ground meat),(spaghetti),0.098254,0.174110,0.039195,0.398915,2.291162,1.0,0.022088,1.373997,0.624943,0.168096,0.272197,0.312015
59,(olive oil),(mineral water),0.065858,0.238368,0.027596,0.419028,1.757904,1.0,0.011898,1.310962,0.461536,0.099759,0.237201,0.267400
51,(ground meat),(mineral water),0.098254,0.238368,0.040928,0.416554,1.747522,1.0,0.017507,1.305401,0.474369,0.138413,0.233952,0.294127
67,(olive oil),(spaghetti),0.065858,0.174110,0.022930,0.348178,1.999758,1.0,0.011464,1.267048,0.535186,0.105651,0.210764,0.239939


Observations:-

Spaghetti → Ground Meat is the strongest rule in the list, having a lift of 2.29.
Soup → Mineral Water is the strongest rule by conviction which makes sense because without the purchase of mineral water, it is less likely that soup will be purchased.
Rules like ground meat → spaghetti and olive oil → spaghetti can be used for targeted marketing, maybe in the form of combos or dedicated sections, etc..

In [12]:
#Find top 2 & bottom 2 antecedent -> consequent rules based on ‘zhang’s metric’. Explain your findings.
display(top5_conviction[['antecedents', 'consequents', 'conviction']])
display(top5_leverage[['antecedents', 'consequents', 'leverage']])
display(top5_lift[['antecedents', 'consequents', 'lift']])

,antecedents,consequents,conviction
62,(soup),(mineral water),1.401255
53,(ground meat),(spaghetti),1.373997
59,(olive oil),(mineral water),1.310962
51,(ground meat),(mineral water),1.305401
67,(olive oil),(spaghetti),1.267048


,antecedents,consequents,leverage
53,(ground meat),(spaghetti),0.022088
52,(spaghetti),(ground meat),0.022088
64,(spaghetti),(mineral water),0.018223
63,(mineral water),(spaghetti),0.018223
50,(mineral water),(ground meat),0.017507


,antecedents,consequents,lift
52,(spaghetti),(ground meat),2.291162
53,(ground meat),(spaghetti),2.291162
67,(olive oil),(spaghetti),1.999758
62,(soup),(mineral water),1.914955
41,(frozen vegetables),(milk),1.910382


In [13]:
display(rules.sort_values('zhangs_metric',ascending=False).head(2))
display(rules.sort_values('zhangs_metric',ascending=True).head(2))


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
52,(spaghetti),(ground meat),0.174110,0.098254,0.039195,0.225115,2.291162,1.0,0.022088,1.163716,0.682343,0.168096,0.140684,0.312015
53,(ground meat),(spaghetti),0.098254,0.174110,0.039195,0.398915,2.291162,1.0,0.022088,1.373997,0.624943,0.168096,0.272197,0.312015


,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
35,(french fries),(mineral water),0.170911,0.238368,0.033729,0.197348,0.827912,1.0,-0.007011,0.948894,-0.200452,0.089812,-0.053858,0.169423
38,(spaghetti),(french fries),0.174110,0.170911,0.027596,0.158499,0.927381,1.0,-0.002161,0.985251,-0.086602,0.086938,-0.014970,0.159983


Observations:-

Spaghetti → Ground Meat and Ground Meat → Spaghetti show the highest Zhang Metric. This suggests that when one of these items is bought, the other's purchase can be reliably predicted.
French Fries → Mineral Water and Spaghetti → French Fries show Zhang Metric less than 0. These rules show slightly negative to weak association. So it is unlikely that if a customer purchases French Fries, they will also get Mineral Water.